source/common/filesystem/filesystem_impl.cc

```diff
@@ -137,23 +137,25 @@ void FileImpl::doWrite(Buffer::Instance& buffer) {
 }
 
 void FileImpl::flushThreadFunc() {
-  std::unique_lock<std::mutex> lock(write_lock_);
 
   while (true) {
+    std::unique_lock<std::mutex> write_lock(write_lock_);
+
     // flush_event_ can be woken up either by large enough flush_buffer or by timer.
     // In case it was timer, flush_buffer_ can be empty.
     while (flush_buffer_.length() == 0 && !flush_thread_exit_) {
-      flush_event_.wait(lock);
+      flush_event_.wait(write_lock);
     }
 
     if (flush_thread_exit_) {
       return;
     }
 
+    std::lock_guard<std::mutex> flush_lock(flush_lock_);
     ASSERT(flush_buffer_.length() > 0);
     about_to_write_buffer_.move(flush_buffer_);
     ASSERT(flush_buffer_.length() == 0);
-    lock.unlock();
+    write_lock.unlock();
```

Comment:
Can we use scoping and RAII for this?

Reply:
Not easily, because this lock was acquired by the condition variable wait returning, and we have an additional RAII lock for flush_lock that needs to stay in scope past this point.  I think trying to do this with scoping would make it more difficult to understand.

Reply:
Actually, I might be able to make this kinda work; I'll play with it and see how it looks.

Reply:
Ok, found a way to do this; it's subjective whether this is more or less clear.  The form is basically:

    std::unique_lock<std::mutex> second_guard;
    {
        std::unique_lock<std::mutex> first_guard(first_lock_);
        second_lock = std::unique_lock<std::mutex>(second_lock_);
        // do stuff
    }
    // Do more stuff

I'm fine with this or the original; @mattklein123 @htuch or others, what's your preference?

Reply:
Ok, I pushed a version with this change.  I can back it out if we don't like it.